[SwiftUI] SwiftBrowser should support a URL CLI argument

aprotyas opened the pull request:
#### 0a0952a23ca4359a04d380456efc0a7693da44db
<pre>
[SwiftUI] SwiftBrowser should support a URL CLI argument
<a href="https://bugs.webkit.org/show_bug.cgi?id=293826">https://bugs.webkit.org/show_bug.cgi?id=293826</a>
<a href="https://rdar.apple.com/152332512">rdar://152332512</a>

Reviewed by Richard Robinson.

This patch imbues SwiftBrowser with the ability to process a `--url`
CLI argument instead of defaulting to webkit.org as a landing page.

* Tools/Scripts/webkitpy/swiftbrowser/run_webkit_app.py:
(main):
* Tools/SwiftBrowser/Source/Extensions/CommandLine+Extras.swift:
(CommandLine.value(for:)):
  Introduce an extension to CommandLine that allows callers to retrieve
  the `--url` argument value with `CommandLine.value(for: &quot;--url&quot;)`.
* Tools/SwiftBrowser/Source/SwiftBrowser.swift:
(SwiftBrowserApp.addProtocolIfNecessary(to:)):
  Defaults to http if a scheme is missing, but accepts about: and data:
  URLs as-is.
(SwiftBrowserApp.body):
* Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/295657@main">https://commits.webkit.org/295657@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/146fe93ffb43d724964238d9007ec2febedba700

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105793 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25544 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15937 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110990 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56389 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107834 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26074 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34047 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80340 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108799 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20438 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95467 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60653 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/105271 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13564 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55828 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89861 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13603 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113839 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32933 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24268 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/89422 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33297 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91698 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89092 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22709 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33964 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11772 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28415 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32858 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32604 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35953 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34202 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Tools/SwiftBrowser/Source/SwiftBrowser.swift:
```diff
@@ -32,11 +32,28 @@ struct SwiftBrowserApp: App {
         } defaultValue: {
-            let url = URL(string: homepage)!
+            var parsedURL: URL? = nil
+            let args = CommandLine.arguments
```
this can be factored out into a `CommandLine+Extras.swift` file and then have a function like

```
extension CommandLine {
    static func value(for flag: String) -> String? {
        guard let flagIndex = arguments.firstIndex(of: flag) else {
            return nil
        }

        let valueIndex = arguments.index(after: flagIndex)
        guard valueIndex < arguments.endIndex else {
            return nil
        }

        return arguments[valueIndex]
    }
}
```

rr-codes commented on Tools/SwiftBrowser/Source/SwiftBrowser.swift:
```diff
@@ -32,11 +32,28 @@ struct SwiftBrowserApp: App {
 
     @State private var mostRecentURL: URL? = nil
 
+    func addProtocolIfNecessary(_ address: String) -> String {
```
this can be `private static`

aprotyas committed:
[SwiftUI] SwiftBrowser should support a URL CLI argument
https://bugs.webkit.org/show_bug.cgi?id=293826
rdar://152332512

Reviewed by Richard Robinson.

This patch imbues SwiftBrowser with the ability to process a `--url`
CLI argument instead of defaulting to webkit.org as a landing page.

* Tools/Scripts/webkitpy/swiftbrowser/run_webkit_app.py:
(main):
* Tools/SwiftBrowser/Source/Extensions/CommandLine+Extras.swift:
(CommandLine.value(for:)):
  Introduce an extension to CommandLine that allows callers to retrieve
  the `--url` argument value with `CommandLine.value(for: "--url")`.
* Tools/SwiftBrowser/Source/SwiftBrowser.swift:
(SwiftBrowserApp.addProtocolIfNecessary(to:)):
  Defaults to http if a scheme is missing, but accepts about: and data:
  URLs as-is.
(SwiftBrowserApp.body):
* Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/295657@main
